codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpProtocolUnitTestGenerator.java

```diff
@@ -0,0 +1,497 @@
+    abstract String unitTestFuncNameFormat();
+
+    /**
+     * Provides the unit test function name's format string arguments.
+     *
+     * @return returns a list of arguments used to format the unitTestFuncNameFormat returned format string.
+     */
+    abstract Object[] unitTestFuncNameArgs();
+
+
+    /**
+     * Hook to provide custom generated code within a test function before test cases are defined.
+     *
+     * @param writer writer to write generated code with.
+     */
+    protected void generateTestSetup(GoWriter writer) {
+        if (hasIdempotencyTokenInputMember()) {
+            writer.write("origRandReader := randReader");
+            writer.write("defer func() { randReader = origRandReader }()");
+            writer.write("randReader = &smithytesting.ByteLoop{}");
+            writer.write("");
+        }
+    }
+
+    /**
+     * Hook to generate the parameter declarations as struct parameters into the test case's struct definition.
+     * Must generate all test case parameters before returning.
+     *
+     * @param writer writer to write generated code with.
+     */
+    abstract void generateTestCaseParams(GoWriter writer);
+
+    /**
+     * Hook to generate all the test case parameters as struct member values for a single test case.
+     * Must generate all test case parameter values before returning.
+     *
+     * @param writer   writer to write generated code with.
+     * @param testCase definition of a single test case.
+     */
+    abstract void generateTestCaseValues(GoWriter writer, T testCase);
+
+    /**
+     * Hook to generate the body of the test that will be invoked for all test cases of this operation. Should not
+     * do any assertions.
+     *
+     * @param writer writer to write generated code with.
+     */
+    abstract void generateTestBody(GoWriter writer);
+
+    /**
+     * Hook to generate the assertions for the operation's test cases. Will be in the same scope as the test body.
+     *
+     * @param writer writer to write generated code with.
+     */
+    abstract void generateTestAssertions(GoWriter writer);
+
+    /**
+     * Generates the test function for the operation using the provided writer.
+     *
+     * @param writer writer to write generated code with.
+     */
+    public void generateTestFunction(GoWriter writer) {
+        writer.addUseImports(SmithyGoDependency.TESTING);
+        writer.openBlock("func " + unitTestFuncNameFormat() + "(t *testing.T) {", "}", unitTestFuncNameArgs(),
+                () -> {
+                    generateTestSetup(writer);
+
+                    writer.write("cases := map[string]struct {");
+                    generateTestCaseParams(writer);
+                    writer.openBlock("}{", "}", () -> {
+                        for (T testCase : testCases) {
+                            testCase.getDocumentation().ifPresent(writer::writeDocs);
+                            writer.openBlock("$S: {", "},", testCase.getId(), () -> {
```

Comment:
Yeah I'd expect the test cases to be stable, and should be able to support test cases running, or generated in any order. Even transversal of the members of the params document can be arbitrary order since the fields are always referred to by name.